### PR TITLE
Enhancement: Speedup `ServiceManager#has()` for non-aliases

### PR DESCRIPTION
**Pull request by fhein**
Currently has() always implies alias resolution. This PR speeds up has() for all requests which are not alias requests. Alias resolution is done after all other resolution options were tried without success. So alias resulotion gets a bit slower now. 

I did this with the discussion around aliases with @Ocramius (see #213, #214, #215) in mind. The intent of this PR is to remove the performance impact alias resolution has on normal resolution. 

The use of aliases seems to be not recommended. If it has no impact on performance any longer, I hope there is a chance to not deprecate that feature. 

The alias feature is problematic only in aggregated service manager configurations, where aliases can supersede inheritated service and factory definitions (even without intent). For all other scenarios the alias feature is in my opinion valuable and worth maintaining.

This PR is covered by existing tests. No need for a new one.

Happy New Year! :)

### CODE REVIEW

**Comment by fhein** [comment]
Btw, the benchmark suite does not run on Windows without applying a PR to phpbench. See https://github.com/phpbench/phpbench/pull/498

File: src/ServiceManager.php [hunk 1]
```diff
@@ -232,20 +232,27 @@ public function build($name, array $options = null)
...
             }
         }
 
+        // If $name is no alias, we are done
+        if (! isset($this->resolvedAliases[$name])) {
+            return false;
+        }
+
+        // Finally check aliases
+        $name = $this->resolvedAliases[$name];
```
Comment (Ocramius) [inline src/ServiceManager.php]: Avoid overriding an existing variable: define a new one instead 👍

File: src/ServiceManager.php [hunk 2]
```diff
@@ -232,20 +232,27 @@ public function build($name, array $options = null)
...
+        return isset($this->services[$name]) || isset($this->factories[$name]);
+
         return false;
```
Comment (Ocramius) [inline src/ServiceManager.php]: Unreachable

**Comment by Ocramius** [comment]
@fhein you couldn't run the benchmarks locally, right?

**Comment by fhein** [comment]
Not at the first time. After applying the mentioned PR to phpbench I could.

**Comment by fhein** [comment]
Introduced $resolvedName to avoid $name reuse. Removed unreachable return.

**Comment by Ocramius** [comment]
Overall good improvement - just need to check the benchmark results when possible. (Can't personally due to being stuck on other work, sorry)

**Comment by fhein** [comment]
@Ocramius: The benchmark suite currently does not test has() at all afaics. The benefits of the short-circuit in has are quite obvious, I believe. Do you think, an explicit benchmark is needed?

The more important case is get() (#215). For #215 I think a benchmark is required. If you agree, I'd appreciate to take care of that and provide the results.

**Comment by Ocramius** [comment]
A benchmark is mostly needed to prevent regressions, not to prove this
patch specifically (although that is still quite interesting)

**Comment by fhein** [comment]
I'll take care of that.

**Comment by fhein** [comment]
@Ocramius: For some reason phpbench's compare exits with nasty errors saying column not found or something. I can provide the dumped xml files, if you want them. 

Here are the results of the performance test for ::has(). First is master, second is this PR. As promised this PR is superior. IN all aspects, not only for services and factories. 

Could you please review the hasBench.php. Just to make sure that I did not mess up something.

Btw, I had to increase revolutions and iterations significantly to get usable results. Maybe this applies to the other tests as well. I'll see. 


	Dumped result to has.master.xml
	suite: 133ec875223eb1c55bb664c57b1a416a780a9429, date: 2018-01-03, stime: 19:48:57
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+
	| benchmark | subject                 | groups | params | revs   | its | mem_peak   | best       | mean       | mode       | worst      | stdev      | rstdev | diff    |
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+
	| HasBench  | benchHasFactory1        |        | []     | 100000 | 100 | 1,383,488b | 0.450000µs | 0.464200µs | 0.460039µs | 0.480000µs | 0.008268µs | 1.78%  | +0.04%  |
	| HasBench  | benchHasInvokable1      |        | []     | 100000 | 100 | 1,383,488b | 0.540000µs | 0.557600µs | 0.559804µs | 0.580000µs | 0.009394µs | 1.68%  | +20.17% |
	| HasBench  | benchHasService1        |        | []     | 100000 | 100 | 1,383,488b | 0.450000µs | 0.464000µs | 0.460098µs | 0.480000µs | 0.008718µs | 1.88%  | 0.00%   |
	| HasBench  | benchHasAlias1          |        | []     | 100000 | 100 | 1,383,480b | 0.550000µs | 0.563800µs | 0.560098µs | 0.590000µs | 0.009673µs | 1.72%  | +21.51% |
	| HasBench  | benchHasRecursiveAlias1 |        | []     | 100000 | 100 | 1,383,488b | 0.570000µs | 0.584100µs | 0.580333µs | 0.610000µs | 0.011233µs | 1.92%  | +25.88% |
	| HasBench  | benchHasRecursiveAlias2 |        | []     | 100000 | 100 | 1,383,488b | 0.570000µs | 0.583500µs | 0.580176µs | 0.610000µs | 0.009421µs | 1.61%  | +25.75% |
	| HasBench  | benchHasAbstractFactory |        | []     | 100000 | 100 | 1,383,488b | 0.800000µs | 0.813400µs | 0.810078µs | 0.850000µs | 0.011851µs | 1.46%  | +75.30% |
	| HasBench  | benchHasNot             |        | []     | 100000 | 100 | 1,383,480b | 0.790000µs | 0.817000µs | 0.812896µs | 0.850000µs | 0.013528µs | 1.66%  | +76.08% |
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+

	Dumped result to has.216.xml
	suite: 133ec87171d0d0aef2a8a01bd50e09d908c639ae, date: 2018-01-03, stime: 19:36:58
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+
	| benchmark | subject                 | groups | params | revs   | its | mem_peak   | best       | mean       | mode       | worst      | stdev      | rstdev | diff    |
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+
	| HasBench  | benchHasFactory1        |        | []     | 100000 | 100 | 1,387,488b | 0.399960µs | 0.415658µs | 0.409999µs | 0.429960µs | 0.007107µs | 1.71%  | 0.00%   |
	| HasBench  | benchHasInvokable1      |        | []     | 100000 | 100 | 1,387,488b | 0.419950µs | 0.430457µs | 0.429816µs | 0.449960µs | 0.009420µs | 2.19%  | +3.56%  |
	| HasBench  | benchHasService1        |        | []     | 100000 | 100 | 1,387,488b | 0.409950µs | 0.419958µs | 0.419934µs | 0.439960µs | 0.008245µs | 1.96%  | +1.03%  |
	| HasBench  | benchHasAlias1          |        | []     | 100000 | 100 | 1,387,480b | 0.419960µs | 0.429996µs | 0.429836µs | 0.450000µs | 0.009274µs | 2.16%  | +3.45%  |
	| HasBench  | benchHasRecursiveAlias1 |        | []     | 100000 | 100 | 1,387,488b | 0.420000µs | 0.435100µs | 0.430098µs | 0.450000µs | 0.008426µs | 1.94%  | +4.68%  |
	| HasBench  | benchHasRecursiveAlias2 |        | []     | 100000 | 100 | 1,387,488b | 0.430000µs | 0.442800µs | 0.440039µs | 0.460000µs | 0.009174µs | 2.07%  | +6.53%  |
	| HasBench  | benchHasAbstractFactory |        | []     | 100000 | 100 | 1,387,488b | 0.690000µs | 0.707800µs | 0.709472µs | 0.740000µs | 0.010352µs | 1.46%  | +70.28% |
	| HasBench  | benchHasNot             |        | []     | 100000 | 100 | 1,387,480b | 0.750000µs | 0.765400µs | 0.760274µs | 0.800000µs | 0.011868µs | 1.55%  | +84.14% |
	+-----------+-------------------------+--------+--------+--------+-----+------------+------------+------------+------------+------------+------------+--------+---------+

**Comment by Ocramius** [comment]
Changed base to `develop`. @fhein please rebase when you have a chance

**Comment by fhein** [comment]
Closed in favour of #220

**Comment by Ocramius** [comment]
Manually applying this patch to `develop`

**Comment by Ocramius** [comment]
Manually applied to `develop` in https://github.com/zendframework/zend-servicemanager/commit/8bf0fbc67fb8350d6927e6ff974e8df60bcd7d0e, but I had to revert changes applied by this patch after introducing a test that verifies that a breaking change in abstract factory usage was introduced.

See https://github.com/zendframework/zend-servicemanager/commit/9063df725c5aa1b15cf0584ecf5b7c70c2874c42

**Pull request closed without merging**